libsplinter/src/biome/refresh_tokens/store/diesel/postgres.rs

```diff
@@ -0,0 +1,131 @@
+// Copyright 2018-2020 Cargill Incorporated
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::biome::{
+    datastore::{
+        models::{NewRefreshToken, RefreshToken, UserModel},
+        schema::{refresh_tokens, splinter_user},
+    },
+    refresh_tokens::store::{RefreshTokenError, RefreshTokenStore},
+};
+use crate::database::ConnectionPool;
+
+use diesel::{
+    dsl::{delete, insert_into, update},
+    prelude::*,
+    result::Error::NotFound,
+};
+
+pub struct SplinterRefreshTokenStore {
+    connection: ConnectionPool,
+}
+
+impl RefreshTokenStore for SplinterRefreshTokenStore {
+    fn add_token(&self, user_id: &str, token: &str) -> Result<(), RefreshTokenError> {
+        let conn = self.connection.get()?;
+
+        splinter_user::table
```

Comment:
This should follow the operations pattern, like other biome stores.

Reply:
I also have serious reservations about implementing a trait for a struct in a different file or module than the one that it's defined in. I think that should be avoided whenever possible, and we should follow this layout whenever possible.

```
struct Foo {}
impl Foo {}
impl Bar for Foo {}
impl Bizz for Foo {}
```

Reply:
The operations pattern allows us to target as broad a slice of diesel-supported databases as possible at the operation level.  For example, we can query both PostgreSQL and sqlite using a single implementation, but inserts may require different implementations, depending on the table structure or insert statements. 

For example, sqlite doesn't support default column values and PostgreSQL does.  In the `UserStore`, this will require two implementations of the `UserStoreAddUserOperation` for each type of connection. 

As for the second question, I'm not sure I understand what's raising the objection.  Can you elaborate with a more concrete example?

Reply:
I issue is with how the operations object is defined [here](https://github.com/Cargill/splinter/blob/master/libsplinter/src/biome/credentials/store/diesel/operations/mod.rs), but its trait implementations are spread across the operations module. For examle,

https://github.com/Cargill/splinter/blob/03e9bfaaba4a4070ad5060628901566ac53b57f3/libsplinter/src/biome/credentials/store/diesel/operations/fetch_credential_by_username.rs#L28

Basically, I think those implementations should be in on module along side the operations definition.

Reply:
The operations pattern is necessary to deal with how Diesel works -- it's complex, but essentially boils down to being able to pick/choose which implementations of operations work for each backend database. For example, we may have 4 of 5 operations work for both, and 1 of 5 different between postgresql and sqlite.

Reply:
re:Organization of operations - follow the currently pattern, it has been carefully considered. If we change it at some point, we will do so across all of libsplinter at once. That discussion can happen outside this PR, and needs to occur after the sqlite implementation has been finished.